File: src/main/java/com/hoi4utils/clauzewitz/script/ScopedPDXScript.java
package main.java.com.hoi4utils.clauzewitz.script;

import main.java.com.hoi4utils.clauzewitz.code.scope.Scope;
import main.java.com.hoi4utils.clauzewitz.code.scope.ScopeType;

import java.util.Collection;

public interface ScopedPDXScript {
    Scope getDefinitionScope();

    default boolean isSupportedInScope(Scope scope) {
        if (supportedScopes().contains(ScopeType.any))
            return true;
        return supportedScopes().contains(scope.targetScopeType());
    }

    Collection<ScopeType> supportedScopes();
}

File: src/main/scala/com/hoi4utils/script/ScopedPDXScript.scala
package com.hoi4utils.script

import com.hoi4utils.hoi4.scope.{Scope, ScopeType}

trait ScopedPDXScript {
  def definitionScope: Option[Scope]

  def isSupportedInScope(scope: Scope): Boolean = {
    if (supportedScopes.exists(_ == ScopeType.any)) return true
    supportedScopes.exists(_ == scope.targetScopeType)
  }

  def supportedScopes: Iterable[ScopeType]
}
